go.mod

```diff
@@ -54,34 +54,34 @@ require (
 	github.com/prometheus/client_model v0.6.1
 	github.com/prometheus/common v0.62.0
 	github.com/robfig/cron/v3 v3.0.1
-	github.com/spf13/cobra v1.8.1
-	github.com/spf13/pflag v1.0.5
+	github.com/spf13/cobra v1.9.1
+	github.com/spf13/pflag v1.0.6
 	github.com/stretchr/testify v1.10.0
 	github.com/vishvananda/netlink v1.3.1-0.20250206174618-62fb240731fa
 	github.com/vishvananda/netns v0.0.4
-	go.etcd.io/etcd/api/v3 v3.5.16
-	go.etcd.io/etcd/client/pkg/v3 v3.5.16
-	go.etcd.io/etcd/client/v3 v3.5.16
+	go.etcd.io/etcd/api/v3 v3.6.0-alpha.0.0.20250326120537-27336678946c
```

Comment:
We have already cut etcd `release-3.6` branch, please try tag `v3.6.0-rc.2` (we will release rc.3 tomorrow).

Reply:
hrmm... seeing an error trying to use that tag:

```
go work sync
go: go.etcd.io/etcd/server/v3@v3.6.0-rc.2 requires
	go.etcd.io/etcd/client/v2@v2.306.0-rc.2: missing go.etcd.io/etcd/client/go.mod and .../v2/go.mod at revision client/v2.306.0-rc.2
```

Reply:
We changed the client/v2 into an internal module https://github.com/etcd-io/etcd/tree/main/client/internal/v2.  Did you see similar issue when use the latest commit in main branch? 

The client/v2 is only used by the v2discovery, which has already been deprecated. We will completely remove it in 3.7.

Reply:
> Did you see similar issue when use the latest commit in main branch?

no, `main` worked

Reply:
> We changed the client/v2 into an internal module

that seems to be breaking use of `go.etcd.io/etcd/server/v3@v3.6.0-rc.2`

Reply:
for go.etcd.io/etcd/server/v3/embed for in-process unit / integration tests

Reply:
But I agree it's a generic problem, we need to sort it out regardless. 

I think we made a mistake to hide the client/v2 (in https://github.com/etcd-io/etcd/pull/15591), which is still used by etcd/server. 

Two possible solutions that I can think of for now:
- either revert the change (do not hide client/v2 in 3.6);
- decommission the v2discovery in release-3.6. But it may break the existing users which are still using the deprecated v2discovery. 

cc @serathius

Reply:
was it known / intentional that go.etcd.io/etcd/server/v3/embed become unusable?

Reply:
> either revert the change (do not hide client/v2 in 3.6)

given how close we are to 3.6, this is simpler / safer

> decommission the v2discovery in release-3.6. But it may break the existing users which are still using the deprecated v2discovery.

this seems good to do long-term, in an orderly way... maybe announce a target removal release, and give a way to opt-into disabling v2discovery in 3.6 or 3.7, and plan to drop in 3.7 or 3.8?

Reply:
We can easily reproduce this problem, 

It seems any applications, which are depending on the `go.etcd.io/etcd/server/v3`, are affected.  :( 

```
$ go get go.etcd.io/etcd/server/v3@v3.6.0-rc.2
go: downloading go.etcd.io/etcd/server/v3 v3.6.0-rc.2
go: go.etcd.io/etcd/server/v3@v3.6.0-rc.2 requires go.etcd.io/etcd/client/v2@v2.306.0-rc.2: missing go.etcd.io/etcd/client/go.mod and .../v2/go.mod at revision client/v2.306.0-rc.2

```

Reply:
Raised https://github.com/etcd-io/etcd/issues/19670

Another blocker for 3.6.0 :(

Reply:
> maybe announce a target removal release

We are planning to remove it in 3.7. All v2 related stuff will gone in 3.7

Reply:
I'm very confused by https://github.com/etcd-io/etcd/blob/v3.6.0-rc.2/server/go.mod#L31 is the confusing line ... no such module exists 

https://github.com/etcd-io/etcd/tree/client/v2.306.0-rc.2/client has no go.mod or v2/go.mod file ... that's required for that module at that tag to exist

Probably best to hoist this into an etcd repo issue

Reply:
> I'm very confused by the dependency at https://github.com/etcd-io/etcd/blob/v3.6.0-rc.2/server/go.mod#L31 ... no such module exists:

It does exist. The problem is that it's internal, so it can't be used by module outside the repo

https://github.com/etcd-io/etcd/blob/27336678946c98bf272460e5dfc54909ecafc63e/client/internal/v2/go.mod#L1

Reply:
`internal` in the *package* means things, but that module does not exist in a way Go can resolve

Reply:
sorry for the inconvenience. One more lesson learned.

Reply:
all good, glad I happened to try to update this and you happened to mention the rc tags were available so we caught this before .0 :)

Reply:
> all good, glad I happened to try to update this and you happened to mention the rc tags were available so we caught this before .0 :)

We will fill the gap in test :) i.e. adding an example application and verify all the new dependencies/tags are resolvable

Reply:
Could we move `go.etcd.io/etcd/client/v2` into `go.etcd.io/etcd/server/internal/client/v2`?

Reply:
> Could we move `go.etcd.io/etcd/client/v2` into `go.etcd.io/etcd/server/internal/client/v2`?

Seems also a valid option. Let's try it out.

Reply:
+1, dropping the client/v2 dep entirely sounds really good

Reply:
Just raised https://github.com/etcd-io/etcd/pull/19673

Reply:
updated to release-3.6 successfully

Reply:
thx for the confirmation.